File: http/http-client-annotation-processor/src/main/java/io/koraframework/http/client/annotation/processor/extension/HttpClientKoraExtension.java
package io.koraframework.http.client.annotation.processor.extension;

import com.palantir.javapoet.CodeBlock;
import com.palantir.javapoet.ParameterizedTypeName;
import com.palantir.javapoet.TypeName;
import org.jspecify.annotations.Nullable;
import io.koraframework.annotation.processor.common.AnnotationUtils;
import io.koraframework.annotation.processor.common.TagUtils;
import io.koraframework.http.client.annotation.processor.HttpClientAnnotationProcessor;
import io.koraframework.http.client.annotation.processor.HttpClientClassNames;
import io.koraframework.http.client.annotation.processor.HttpClientUtils;
import io.koraframework.kora.app.annotation.processor.extension.ExtensionResult;
import io.koraframework.kora.app.annotation.processor.extension.KoraExtension;

import javax.annotation.processing.ProcessingEnvironment;
import javax.annotation.processing.RoundEnvironment;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class HttpClientKoraExtension implements KoraExtension {
    private final Elements elements;
    private final Types types;
    private final HttpClientAnnotationProcessor processor;

    public HttpClientKoraExtension(ProcessingEnvironment processingEnvironment) {
        this.elements = processingEnvironment.getElementUtils();
        this.types = processingEnvironment.getTypeUtils();
        this.processor = new HttpClientAnnotationProcessor();
        this.processor.init(processingEnvironment);
    }

    @Nullable
    @Override
    public KoraExtensionDependencyGenerator getDependencyGenerator(RoundEnvironment roundEnvironment, TypeMirror typeMirror, String tag) {
        var typeName = TypeName.get(typeMirror);
        if (typeName instanceof ParameterizedTypeName mapper && mapper.rawType().equals(HttpClientClassNames.httpClientResponseMapper)) {
            if (mapper.typeArguments().getFirst() instanceof ParameterizedTypeName either && either.rawType().equals(HttpClientClassNames.either)) {
                var mapperTypeMirror = (DeclaredType) typeMirror;
                var eitherTypeMirror = (DeclaredType) mapperTypeMirror.getTypeArguments().getFirst();
                var responseMapperElement = this.elements.getTypeElement(HttpClientClassNames.httpClientResponseMapper.canonicalName());
                var eitherMapperElement = this.elements.getTypeElement(HttpClientClassNames.httpClientEitherResponseMapper.canonicalName());
                var successType = eitherTypeMirror.getTypeArguments().get(0);
                var errorType = eitherTypeMirror.getTypeArguments().get(1);
                var successTag = tag != null ? tag : TagUtils.parseTagValue(successType);
                var errorTag = tag != null ? tag : TagUtils.parseTagValue(errorType);
                return () -> new ExtensionResult.CodeBlockResult(
                    eitherMapperElement,
                    dependencies -> CodeBlock.of("new $T<>($L)", HttpClientClassNames.httpClientEitherResponseMapper, dependencies),
                    mapperTypeMirror,
                    tag,
                    List.of(
                        this.types.getDeclaredType(responseMapperElement, successType),
                        this.types.getDeclaredType(responseMapperElement, errorType)
                    ),
                    List.of(successTag, errorTag)
                );
            }
            if (mapper.typeArguments().getFirst() instanceof ParameterizedTypeName entity && entity.rawType().equals(HttpClientClassNames.httpResponseEntity)) {
                var mapperTypeMirror = (DeclaredType) typeMirror;
                var entityTypeMirror = (DeclaredType) mapperTypeMirror.getTypeArguments().getFirst();
                var responseMapperElement = this.elements.getTypeElement(HttpClientClassNames.httpClientResponseMapper.canonicalName());
                var responseEntityMapperElement = this.elements.getTypeElement(HttpClientClassNames.httpClientResponseEntityMapper.canonicalName());
                var responseType = entityTypeMirror.getTypeArguments().getFirst();
                if (TypeName.get(responseType) instanceof ParameterizedTypeName eitherResponse && eitherResponse.rawType().equals(HttpClientClassNames.either)) {
                    var eitherTypeMirror = (DeclaredType) responseType;
                    var successType = eitherTypeMirror.getTypeArguments().get(0);
                    var errorType = eitherTypeMirror.getTypeArguments().get(1);
                    var successTag = tag != null ? tag : TagUtils.parseTagValue(successType);
                    var errorTag = tag != null ? tag : TagUtils.parseTagValue(errorType);
                    return () -> new ExtensionResult.CodeBlockResult(
                        responseEntityMapperElement,
                        dependencies -> CodeBlock.of("new $T<>(new $T<>($L))", HttpClientClassNames.httpClientResponseEntityMapper, HttpClientClassNames.httpClientEitherResponseMapper, dependencies),
                        mapperTypeMirror,
                        tag,
                        List.of(
                            this.types.getDeclaredType(responseMapperElement, successType),
                            this.types.getDeclaredType(responseMapperElement, errorType)
                        ),
                        List.of(successTag, errorTag)
                    );
                }
                var tags = new ArrayList<String>();
                tags.add(tag);
                return () -> new ExtensionResult.CodeBlockResult(
                    responseEntityMapperElement,
                    dependencies -> CodeBlock.of("new $T<>($L)", HttpClientClassNames.httpClientResponseEntityMapper, dependencies),
                    mapperTypeMirror,
                    tag,
                    List.of(this.types.getDeclaredType(responseMapperElement, responseType)),
                    tags
                );
            }
            if (HttpClientClassNames.json.canonicalName().equals(tag)) {
                var mapperTypeMirror = (DeclaredType) typeMirror;
                var responseType = mapperTypeMirror.getTypeArguments().getFirst();
                var jsonReaderElement = this.elements.getTypeElement(HttpClientClassNames.jsonReader.canonicalName());
                var jsonMapperElement = this.elements.getTypeElement(HttpClientClassNames.jsonHttpClientResponseMapper.canonicalName());
                return () -> new ExtensionResult.CodeBlockResult(
                    jsonMapperElement,
                    dependencies -> CodeBlock.of("new $T<>($L)", HttpClientClassNames.jsonHttpClientResponseMapper, dependencies),
                    mapperTypeMirror,
                    tag,
                    List.of(this.types.getDeclaredType(jsonReaderElement, responseType)),
                    Collections.singletonList(null)
                );
            }
            return null;
        }

        if (tag != null) return null;
        var element = this.types.asElement(typeMirror);
        if (element == null || element.getKind() != ElementKind.INTERFACE) {
            return null;
        }
        var annotation = AnnotationUtils.findAnnotation(element, HttpClientClassNames.httpClientAnnotation);
        if (annotation == null) {
            return null;
        }
        var typeElement = (TypeElement) element;
        var implName = HttpClientUtils.clientName(typeElement);
        return KoraExtensionDependencyGenerator.generatedFromWithName(elements, element, implName);
    }
}

File: http/http-client-symbol-processor/src/main/kotlin/io/koraframework/http/client/symbol/processor/extension/HttpClientKoraExtension.kt
package io.koraframework.http.client.symbol.processor.extension

import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.ClassKind
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSType
import com.google.devtools.ksp.symbol.Variance
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.ParameterizedTypeName
import com.squareup.kotlinpoet.ksp.toTypeName
import io.koraframework.http.client.symbol.processor.HttpClientClassNames
import io.koraframework.http.client.symbol.processor.HttpClientClassNames.httpClientAnnotation
import io.koraframework.http.client.symbol.processor.clientName
import io.koraframework.kora.app.ksp.KoraAppUtils.findSinglePublicConstructor
import io.koraframework.kora.app.ksp.extension.ExtensionResult
import io.koraframework.kora.app.ksp.extension.KoraExtension
import io.koraframework.ksp.common.AnnotationUtils.findAnnotation
import io.koraframework.ksp.common.KspCommonUtils.getClassDeclarationByName
import io.koraframework.ksp.common.TagUtils.parseTag

class HttpClientKoraExtension : KoraExtension {
    override fun getDependencyGenerator(resolver: Resolver, type: KSType, tag: String?): (() -> ExtensionResult)? {
        val typeName = type.toTypeName()
        if (typeName is ParameterizedTypeName && typeName.rawType == HttpClientClassNames.httpClientResponseMapper) {
            val rsTypeName = typeName.typeArguments.first()
            val rsType = type.arguments.first().type!!.resolve()
            if (rsTypeName is ParameterizedTypeName && rsTypeName.rawType == HttpClientClassNames.either) {
                val responseMapperElement = resolver.getClassDeclarationByName(HttpClientClassNames.httpClientResponseMapper)!!
                val eitherMapperElement = resolver.getClassDeclarationByName(HttpClientClassNames.httpClientEitherResponseMapper)!!
                val successType = rsType.arguments[0].type!!.resolve().makeNotNullable()
                val errorType = rsType.arguments[1].type!!.resolve().makeNotNullable()
                val successTag = tag ?: successType.parseTag()
                val errorTag = tag ?: errorType.parseTag()
                return {
                    val successMapperType = responseMapperElement.asStarProjectedType().replace(listOf(resolver.getTypeArgument(resolver.createKSTypeReferenceFromKSType(successType), Variance.INVARIANT)))
                    val errorMapperType = responseMapperElement.asStarProjectedType().replace(listOf(resolver.getTypeArgument(resolver.createKSTypeReferenceFromKSType(errorType), Variance.INVARIANT)))
                    ExtensionResult.CodeBlockResult(
                        (eitherMapperElement.primaryConstructor ?: eitherMapperElement.findSinglePublicConstructor()),
                        { dependencies -> CodeBlock.of("%T(%L)", HttpClientClassNames.httpClientEitherResponseMapper, dependencies) },
                        type,
                        tag,
                        listOf(successMapperType, errorMapperType),
                        listOf(successTag, errorTag)
                    )
                }
            }
            if (rsTypeName is ParameterizedTypeName && rsTypeName.rawType == HttpClientClassNames.httpResponseEntity) {
                val responseMapperElement = resolver.getClassDeclarationByName(HttpClientClassNames.httpClientResponseMapper)!!
                val responseEntityMapperElement = resolver.getClassDeclarationByName(HttpClientClassNames.httpClientResponseEntityMapper)!!
                val responseType = rsType.arguments.first().type!!.resolve().makeNotNullable()
                val responseTypeName = rsTypeName.typeArguments.first()
                if (responseTypeName is ParameterizedTypeName && responseTypeName.rawType == HttpClientClassNames.either) {
                    val successType = responseType.arguments[0].type!!.resolve().makeNotNullable()
                    val errorType = responseType.arguments[1].type!!.resolve().makeNotNullable()
                    val successTag = tag ?: successType.parseTag()
                    val errorTag = tag ?: errorType.parseTag()
                    return {
                        val successMapperType = responseMapperElement.asStarProjectedType().replace(listOf(resolver.getTypeArgument(resolver.createKSTypeReferenceFromKSType(successType), Variance.INVARIANT)))
                        val errorMapperType = responseMapperElement.asStarProjectedType().replace(listOf(resolver.getTypeArgument(resolver.createKSTypeReferenceFromKSType(errorType), Variance.INVARIANT)))
                        ExtensionResult.CodeBlockResult(
                            (responseEntityMapperElement.primaryConstructor ?: responseEntityMapperElement.findSinglePublicConstructor()),
                            { dependencies -> CodeBlock.of("%T(%T(%L))", HttpClientClassNames.httpClientResponseEntityMapper, HttpClientClassNames.httpClientEitherResponseMapper, dependencies) },
                            type,
                            tag,
                            listOf(successMapperType, errorMapperType),
                            listOf(successTag, errorTag)
                        )
                    }
                }
                val tags = ArrayList<String?>()
                tags.add(tag)
                return {
                    val delegateType = responseMapperElement.asStarProjectedType().replace(listOf(resolver.getTypeArgument(resolver.createKSTypeReferenceFromKSType(responseType), Variance.INVARIANT)))
                    ExtensionResult.CodeBlockResult(
                        (responseEntityMapperElement.primaryConstructor ?: responseEntityMapperElement.findSinglePublicConstructor()),
                        { dependencies -> CodeBlock.of("%T(%L)", HttpClientClassNames.httpClientResponseEntityMapper, dependencies) },
                        type,
                        tag,
                        listOf(delegateType),
                        tags
                    )
                }
            }
            if (tag == HttpClientClassNames.json.canonicalName) {
                val jsonReaderElement = resolver.getClassDeclarationByName(HttpClientClassNames.jsonReader)!!
                val jsonMapperElement = resolver.getClassDeclarationByName(HttpClientClassNames.jsonHttpClientResponseMapper)!!
                return {
                    val jsonReaderType = jsonReaderElement.asStarProjectedType().replace(listOf(resolver.getTypeArgument(resolver.createKSTypeReferenceFromKSType(rsType), Variance.INVARIANT)))
                    ExtensionResult.CodeBlockResult(
                        (jsonMapperElement.primaryConstructor ?: jsonMapperElement.findSinglePublicConstructor()),
                        { dependencies -> CodeBlock.of("%T(%L)", HttpClientClassNames.jsonHttpClientResponseMapper, dependencies) },
                        type,
                        tag,
                        listOf(jsonReaderType),
                        listOf(null)
                    )
                }
            }
            return null
        }


        if (tag != null) return null
        val declaration = type.declaration
        if (declaration !is KSClassDeclaration || declaration.classKind != ClassKind.INTERFACE) {
            return null
        }
        if (declaration.findAnnotation(httpClientAnnotation) == null) {
            return null
        }
        return generatedByProcessorWithName(resolver, declaration, declaration.clientName())
    }
}
